File: tests/fixtures/open_redirect/javascript/safe_redirect.js
// Safe: req.query.next routed through `validateRedirectUrl` allowlist
// before being passed to res.redirect.
function validateRedirectUrl(raw) {
    return raw.startsWith('/') ? raw : '/';
}

function handler(req, res) {
    const target = req.query.next;
    const safe = validateRedirectUrl(target);
    res.redirect(safe);
}

module.exports = handler;

File: tests/fixtures/open_redirect/java/SafeRedirect.java
// Safe: request param routed through `validateRedirectUrl` allowlist
// before being passed to sendRedirect.
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

public class SafeRedirect {
    public static String validateRedirectUrl(String raw) {
        return raw != null && raw.startsWith("/") ? raw : "/";
    }

    public void handle(HttpServletRequest req, HttpServletResponse res) throws IOException {
        String target = req.getParameter("next");
        String safe = validateRedirectUrl(target);
        res.sendRedirect(safe);
    }
}
